torch/custom_class.h

```diff
@@ -101,9 +102,22 @@ class class_ {
     defineMethod<void>("__init__", std::move(func));
     return *this;
   }
-  template <typename Func>
+  template <
+      typename Method,
+      std::enable_if_t<
+          !c10::guts::is_stateless_lambda<std::decay_t<Method>>::value,
+          bool> = false>
+  class_& def(std::string name, Method m) {
+    auto res = def_(name, m, detail::args_t<decltype(m)>{});
+    return *this;
+  }
+  template <
+      typename Func,
+      std::enable_if_t<
+          c10::guts::is_stateless_lambda<std::decay_t<Func>>::value,
+          bool> = false>
   class_& def(std::string name, Func f) {
-    auto res = def_(name, f, detail::args_t<decltype(f)>{});
+    auto res = def_(name, f, detail::args_t<decltype(&Func::operator())>{});
```

Comment:
also move here. Also, why does this need a special case? It should be possible to write this and the !is_stateless_lambda in one implementation.